### PR TITLE
a

### PR DESCRIPTION
**Pull request by stuartwilkes**
### Reasons for making this change

As described [here](https://github.com/rjsf-team/react-jsonschema-form/issues/2752), if an allOf block is present at multiple levels within a schema, and the nested allOf block implements multiple if/then/else statements, the nested dependency is not correctly rendered.

This PR takes the patch described in the linked thread and implements it within the library to fix the issue as described. It also provides a test for the issue.

The current behaviour can be seen as failing in [this playground](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJBbmltYWwiOiJDYXQiLCJUYWlsIjoiTG9uZyJ9LCJzY2hlbWEiOnsidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsiQW5pbWFsIjp7ImRlZmF1bHQiOiJDYXQiLCJlbnVtIjpbIkNhdCIsIkRvZyJdLCJ0aXRsZSI6IkFuaW1hbCIsInR5cGUiOiJzdHJpbmcifX0sImFsbE9mIjpbeyJpZiI6eyJyZXF1aXJlZCI6WyJBbmltYWwiXSwicHJvcGVydGllcyI6eyJBbmltYWwiOnsiY29uc3QiOiJDYXQifX19LCJ0aGVuIjp7InByb3BlcnRpZXMiOnsiVGFpbCI6eyJkZWZhdWx0IjoiTG9uZyIsImVudW0iOlsiTG9uZyIsIlNob3J0IiwiTm9uZSJdLCJ0aXRsZSI6IlRhaWwgbGVuZ3RoIiwidHlwZSI6InN0cmluZyJ9fSwicmVxdWlyZWQiOlsiVGFpbCJdfX0seyJpZiI6eyJyZXF1aXJlZCI6WyJBbmltYWwiXSwicHJvcGVydGllcyI6eyJBbmltYWwiOnsiY29uc3QiOiJEb2cifX19LCJ0aGVuIjp7InByb3BlcnRpZXMiOnsiQnJlZWQiOnsidGl0bGUiOiJCcmVlZCIsInByb3BlcnRpZXMiOnsiQnJlZWROYW1lIjp7ImRlZmF1bHQiOiJBbHNhdGlhbiIsImVudW0iOlsiQWxzYXRpYW4iLCJEYWxtYXRpb24iXSwidGl0bGUiOiJCcmVlZCBuYW1lIiwidHlwZSI6InN0cmluZyJ9fSwiYWxsT2YiOlt7ImlmIjp7InJlcXVpcmVkIjpbIkJyZWVkTmFtZSJdLCJwcm9wZXJ0aWVzIjp7IkJyZWVkTmFtZSI6eyJjb25zdCI6IkFsc2F0aWFuIn19fSwidGhlbiI6eyJwcm9wZXJ0aWVzIjp7IkZ1ciI6eyJkZWZhdWx0IjoiYnJvd24iLCJlbnVtIjpbImJsYWNrIiwiYnJvd24iXSwidGl0bGUiOiJGdXIiLCJ0eXBlIjoic3RyaW5nIn19LCJyZXF1aXJlZCI6WyJGdXIiXX19LHsiaWYiOnsicmVxdWlyZWQiOlsiQnJlZWROYW1lIl0sInByb3BlcnRpZXMiOnsiQnJlZWROYW1lIjp7ImNvbnN0IjoiRGFsbWF0aW9uIn19fSwidGhlbiI6eyJwcm9wZXJ0aWVzIjp7IlNwb3RzIjp7ImRlZmF1bHQiOiJzbWFsbCIsImVudW0iOlsibGFyZ2UiLCJzbWFsbCJdLCJ0aXRsZSI6IlNwb3RzIiwidHlwZSI6InN0cmluZyJ9fSwicmVxdWlyZWQiOlsiU3BvdHMiXX19XSwicmVxdWlyZWQiOlsiQnJlZWROYW1lIl19fX19XSwicmVxdWlyZWQiOlsiQW5pbWFsIl19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoiZGVmYXVsdCIsImxpdmVTZXR0aW5ncyI6eyJ2YWxpZGF0ZSI6ZmFsc2UsImRpc2FibGUiOmZhbHNlLCJyZWFkb25seSI6ZmFsc2UsIm9taXRFeHRyYURhdGEiOmZhbHNlLCJsaXZlT21pdCI6ZmFsc2V9fQ==)

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/master/CHANGELOG.md) with a description of the PR
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/core/src/utils.js [hunk 1]
```diff
@@ -767,6 +767,32 @@ export function retrieveSchema(schema, rootSchema = {}, formData = {}) {
...
+        rootSchema,
+        propData
+      );
+
+      properties[propName] = resolvedPropSchema;
+
+      if (
+        propSchema !== resolvedPropSchema &&
+        resolvedSchema.properties !== properties &&
+        properties !== null
```
Comment (AlimovSV) [inline packages/core/src/utils.js:789]: I think `properties !== null` check is redundand, isn't?

Comment (stuartwilkes) [reply packages/core/src/utils.js]: Thanks, you're right - this was a holdover from when I was addressing an issue that caused the patch to fail [this test](https://github.com/rjsf-team/react-jsonschema-form/blob/d87a88b99c1e9f793b414cb32d412ea9b59293ec/packages/core/test/utils_test.js#L1096)

File: packages/core/test/utils_test.js [hunk 2]
```diff
@@ -2540,6 +2540,282 @@ describe("utils", () => {
...
+              default: "Long",
+              enum: ["Long", "Short", "None"],
+              title: "Tail length",
+              type: "string",
+            },
+          },
+          required: ["Animal", "Tail"],
+        });
+      });
+      it("should resolve multiple conditions in nested allOf blocks", () => {
```
Comment (nickgros) [inline packages/core/test/utils_test.js:2674]: Can we change the test names to capture what the difference is between the tests, and more importantly so we don't have two tests with the same name?

File: packages/core/src/utils.js [hunk 3]
```diff
@@ -767,6 +767,31 @@ export function retrieveSchema(schema, rootSchema = {}, formData = {}) {
     return resolveCondition(schema, rootSchema, formData);
   }
 
+  if (resolvedSchema.properties) {
+    const properties = {};
+
+    Object.entries(resolvedSchema.properties).forEach(entries => {
+      const propName = entries[0];
+      const propSchema = entries[1];
```
Comment (nickgros) [inline packages/core/src/utils.js:775]: Is it possible that a resolved schema will have additional properties such that this ordering doesn't hold? Seems like it would make more sense to filter

File: packages/core/src/utils.js [hunk 4]
```diff
@@ -767,6 +767,31 @@ export function retrieveSchema(schema, rootSchema = {}, formData = {}) {
     return resolveCondition(schema, rootSchema, formData);
   }
 
+  if (resolvedSchema.properties) {
```
Comment (nickgros) [inline packages/core/src/utils.js:770]: Could you add a comment clarifying why we have to do this new step, and what exactly we're doing? It's hard to follow exactly what's going on.

File: packages/core/src/utils.js [hunk 5]
```diff
@@ -767,6 +767,31 @@ export function retrieveSchema(schema, rootSchema = {}, formData = {}) {
...
+        propSchema,
+        rootSchema,
+        propData
+      );
+
+      properties[propName] = resolvedPropSchema;
+
+      if (
+        propSchema !== resolvedPropSchema &&
+        resolvedSchema.properties !== properties
```
Comment (nickgros) [inline packages/core/src/utils.js:788]: It looks like `properties` will always be a different object than `resolvedSchema.properties`, since you declared a new `properties` object on line 771. So I think this would always be true.

File: packages/core/test/utils_test.js (hunk 2 above)
Comment (nickgros) [reply packages/core/test/utils_test.js:2674]: Follow up question: are both tests necessary? Do they capture separate edge cases that we effectively handle differently?

File: packages/core/src/utils.js (hunk 5 above)
Comment (AlimovSV) [reply packages/core/src/utils.js:788]: No, this is possible ONLY when the recursive call of retrieveSchema returns a modified schema for the specific property. Otherwise resolvedSchema will be returned unchanged.

File: packages/core/src/utils.js (hunk 3 above)
Comment (AlimovSV) [reply packages/core/src/utils.js:775]: Could you provide the schema which is not handled correctly by this code? For better understanding the case you are talking about. I use this code in the project with large and complex schemas with nested if/then/else so if you see the case which will not work - tell me please.

File: packages/core/test/utils_test.js (hunk 2 above)
Comment (stuartwilkes) [reply packages/core/test/utils_test.js]: Thanks, you're right, we don't need both these tests, only the second one, which captures the failing behaviour in the current live code and where the new code addresses this issue. Removed.

File: packages/core/src/utils.js (hunk 3 above)
Comment (stuartwilkes) [reply packages/core/src/utils.js:775]: I believe as we're using `entries` here, and the properties will always be objects representing the fields, index 0 will always be `propName` and index 1 the subschema ?

Comment (AlimovSV) [reply packages/core/src/utils.js:775]: Sure, but I guess @nickgros is talking about something like `.additionalProperties`.

Comment (nickgros) [reply packages/core/src/utils.js:775]: I see now, I think I was misunderstanding the code. This makes sense to me. Thanks for the clarification

File: packages/core/src/utils.js (hunk 5 above)
Comment (AlimovSV) [reply packages/core/src/utils.js:788]: The idea of this check is to reduce creation of shallow copy for the retrieved schemas. So if a schema isn't computed schema (doesn't have any if/then/else) it will be returned unchanged (and reference equality check will be enough for this condition)

File: packages/core/src/utils.js (hunk 4 above)
Comment (AlimovSV) [reply packages/core/src/utils.js:770]: finding the right words is not my best side :) I like your description

File: packages/core/src/utils.js (hunk 5 above)
Comment (nickgros) [reply packages/core/src/utils.js:790]: Sorry if I'm being unclear, but essentially I don't think you will ever have referential equality between `resolvedSchema.properties` and `properties` because `properties` is always a new object.

Comment (AlimovSV) [reply packages/core/src/utils.js:790]: No, let me explain. Before loop we initialize `properties` variable as a new object, you right. And during the iteration (first time when we meet `propSchema !== resolvedPropSchema`) the condition `resolvedSchema.properties !== properties` will be truthy, you right too. But after we make a shallow copy of the schema with new properties container `resolvedSchema = { ...resolvedSchema, properties };` during any next iteration `resolvedSchema.properties !== properties` will be falsy and we will not make shallow copy. The bonuse is: when resolvedSchema doesn't contain properties with computed schema we do not make shallow copy at all and return the unmodified object.

Comment (nickgros) [reply packages/core/src/utils.js:790]: 🤦‍♂️ I finally see it now! Thank you for this explanation (and for bearing with me)!

**Review by nickgros — APPROVED** [review]
This looks good to me if tests are passing. I'd like another set of eyes since our schema resolution utils are so complicated

**Review by jacqueswho — APPROVED** [review]

**Review by AlimovSV — APPROVED** [review]
